code/extcon.c

```diff
@@ -0,0 +1,295 @@
+  } MPS_ARGS_END(args);
+
+  printf("Initial reservation %"PRIuLONGEST".\n", (ulongest_t)mps_arena_reserved(arena));
+
+  die(dylan_fmt(&obj_fmt, arena), "dylan_fmt()");
+
+  /* Create new pool */
+  MPS_ARGS_BEGIN(args) {
+    MPS_ARGS_ADD(args, MPS_KEY_FORMAT, obj_fmt);
+    die(mps_pool_create_k(&obj_pool, arena, mps_class_amcz(), args),
+      "mps_pool_create_k");
+  } MPS_ARGS_END(args);
+
+  /* Register thread */
+  die(mps_thread_reg(&thread, arena), "Thread reg");
+
+  /* Register stack roots */
+  /* Since this testbench is currently not using a stack root, #IF 0 this out */
+  testlib_unused(cold_stack_end);
+  testlib_unused(stack_root);
+#if 0
+  die(mps_root_create_thread(&stack_root, arena, thread, cold_stack_end), "Create Stack root");
+#endif
+
+  /* Register ambiguous array of object roots. */
+  die(mps_root_create_area(&testobj_root, arena,
+                            mps_rank_ambig(), (mps_rm_t)0,
+                            &testobj[0], &testobj[N_TESTOBJ],
+                            mps_scan_area, NULL),
+      "root_create_area(testobj)");
+
+  /* Create allocation point */
+  die(mps_ap_create_k(&obj_ap, obj_pool, mps_args_none), "Create Allocation point");
+
+  mps_message_type_enable(arena, mps_message_type_gc_start());
+  mps_message_type_enable(arena, mps_message_type_gc());
+
+  /* Allocate objects and force arena extension */
+  for (i = 0; i < N_TESTOBJ; i++) {
+
+    die(make_dylan_vector(&testobj[i], obj_ap, N_SLOT_TESTOBJ), "make_dylan_vector"); 
+
+    printf("Object %d committed.  "
+           "Arena reserved: %"PRIuLONGEST".\n",
+           i,
+           (ulongest_t)mps_arena_reserved(arena));
+
+    print_messages();
+  }
+
+  /* overwrite all the references to the objects*/
+  for (i = 0; i < N_TESTOBJ; i++) {
+
+    /* bonus test of mps_addr_object */
+#if 0 /* Comment this out until mps_addr_object becomes available. */
+    mps_addr_t out;
+    Insist(N_TESTOBJ <= N_INT_TESTOBJ);
+
+    /* use "i" to as a convenient way to generate different interior pointers
+       To guarantee the i index will give us an interior pointer the number of test
+       objects must be <= the number of integers in each object */
+    Insist(N_TESTOBJ <= N_INT_TESTOBJ);
+    die(mps_addr_object(&out, arena, &(testobj[i])->int_array[i]), "Address object");
+
+    Insist(out == testobj[i]);
+
+    /* end piggy back testbench */
+#endif
+
+    /* now overwrite the ref */
+    testobj[i] = (mps_word_t)NULL;
+
+    print_messages();
+  }
+
+  /* Collect */
+  mps_arena_collect(arena);
+
+  print_messages();
+
+  printf("Arena extended %d times\n", n_extend);
+  printf("Arena contracted %d times\n", n_contract);
+
+  /* Clean up */
+  mps_root_destroy(testobj_root);
+  /* mps_root_destroy(stack_root);*/
+  mps_thread_dereg(thread);
+  mps_ap_destroy(obj_ap);
+  mps_pool_destroy(obj_pool);
+  mps_fmt_destroy(obj_fmt);
+  mps_arena_destroy(arena);
+#if 0
```

Comment:
m. Every #if'd out block must have an explanation.

Reply:
Fix - explanation added in comments